lib/wallaroo/core/common/state_step_creator.pony

```diff
@@ -0,0 +1,286 @@
+use "collections"
+use "itertools"
+use "wallaroo/core/boundary"
+use "wallaroo/core/initialization"
+use "wallaroo/core/metrics"
+use "wallaroo/core/topology"
+use "wallaroo/ent/recovery"
+use "wallaroo/ent/router_registry"
+use "wallaroo_labs/mort"
+
+class KeyToStepInfo[Info: Any val]
+  let _info: Map[String, Map[Key, Info]]
+
+  new create() =>
+    _info = _info.create()
+
+  fun apply(state_name: String, key: box->Key!): this->Info ? =>
+    _info(state_name)?(key)?
+
+  fun ref add(state_name: String, key: Key, info: Info^) =>
+    try
+      _info.insert_if_absent(state_name, Map[Key, Info])?(key) = info
+    end
+
+  fun contains(state_name: String, key: Key): Bool =>
+    try
```

Comment:
if _map.contains(state_name) then
  try
    let inner_m = _map(state_name)? 
  else
     Unreachable()
  end
  inner_m.contains(key)
else
  false
end

Reply:
This could live in a primitive with two type args.

Reply:
done